jujugui/static/gui/src/app/components/deployment-flow/sshkey/sshkey.js

```diff
@@ -30,22 +15,298 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-    @method _onSSHKeyInputBlur
-    @param {Object} evt The blur event.
+    @param {Object} evt The keyboard event.
   */
-  _onSSHKeyInputBlur(evt) {
-    const key = this.refs.sshKey.getValue();
-    this.props.setSSHKey(key);
+  _onKeyUp(evt) {
+    if (evt.which === 13) {
+      this._handleAddMoreKeys(null);
+    }
+    const hasValue = (this.refs.sshKey && this.refs.sshKey.getValue()) ||
+      (this.refs.githubUsername && this.refs.githubUsername.getValue());
+    this.setState({
+      buttonDisabled: hasValue ? false : true
+    });
   }
 
   /**
-    Render the component.
+    Split a Manual Key into its parts.
 
-    @method render
+    @param {String} sshkey
+   */
+  _splitKey(sshKey) {
+    if (sshKey.indexOf(' ') > -1) {
+      const splitKey = sshKey.split(' ');
+      if (splitKey.length >= 2) {
+        return {
+          'body': splitKey[1],
+          'text': sshKey,
+          'type': splitKey[0],
+          'id': 0 // To conform to the schema of the github api
+        };
+      }
+    }
+    this.setState({error: (<span>Key is invalid. Please
+    ensure you have copied the key correctly.</span>)});
+    return false;
+  }
+
+  /**
+    Check if an sshkey exists to avoid duplication.
+
+    @param {Object} key The key object to check.
+    @return {Boolean}
   */
+  _keyExists(key) {
+    return this.state.SSHkeys.
+      findIndex(knownKey => knownKey.body === key.body) !== -1;
+  }
+
+  /**
+    Handle adding github username.
+
+    @param {String} error The error returned.
+    @param {Array} keys The returned keys.
+  */
+  _addGithubKeysCallback(error, keys) {
+    if (error) {
+      console.error(error);
+      this.setState({error: error});
+      return;
+    }
+
+    if (keys.length === 0) {
+      this.setState({error:
+        (<span>No keys found. <a className="link"
+          href="https://github.com/settings/keys" target="_blank">
+          Create an SSH Key</a>.</span>)});
+      return;
+    }
+
+    let SSHkeys = this.state.SSHkeys;
+    keys.forEach(key => {
+      if (!this._keyExists(key)) {
+        SSHkeys.push(key);
+      }
+    });
+    this.props.setSSHKey(SSHkeys[0].text);
+    this.setState({SSHkeys: SSHkeys, buttonDisabled: true});
+    this.refs.githubUsername.setValue(null);
+    this.refs.githubUsername.focus();
+  }
+
+  /**
+    Handle clicking AddMoreKeys.
+  */
+  _handleAddMoreKeys() {
+    if (this.state.buttonDisabled) {
+      return;
+    }
+    const source = this.refs.sshSource.getValue();
+    this.setState({error: null});
+    if (source === 'github') {
+      const githubUsername = this.refs.githubUsername.getValue();
+      this.props.githubSSHKeys(
+        new yui.juju.environments.web.WebHandler,
+        githubUsername,
+        this._addGithubKeysCallback.bind(this)
+      );
+    } else if (source === 'manual') {
+      const manualKey = this.refs.sshKey.getValue();
+      const key = this._splitKey(manualKey);
+      if (key) {
```

Comment:
if no key we should indicate the issue to the user. I suppose this is done in `_splitKey` so maybe `_validateAndSplitKey` is more appropriate?